internal/agent/agent.go

```diff
@@ -0,0 +1,102 @@
+package agent
+
+import (
+	"fmt"
+	"log"
+	"math/rand/v2"
+	"net/http"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type MetricsCollector struct {
+	mx                  sync.Mutex
+	GaugeMetrics        map[string]interface{}
+	PollInterval        int64
+	ReportCountInterval int64
+	Addr                string
+	PollCount           int64
+}
+
+func New(flagRunAddr string, reportInterval, pollInterval int64) *MetricsCollector {
+	return &MetricsCollector{
+		GaugeMetrics:        map[string]interface{}{},
+		PollInterval:        pollInterval,
+		ReportCountInterval: reportInterval,
+		Addr:                flagRunAddr,
+		PollCount:           0,
+	}
+}
+func (m *MetricsCollector) UpdateMetrics() {
+	m.PollCount++
+	var memStat runtime.MemStats
+	runtime.ReadMemStats(&memStat)
+	m.mx.Lock()
+	defer m.mx.Unlock()
+	m.GaugeMetrics["Alloc"] = memStat.Alloc
+	m.GaugeMetrics["BuckHashSys"] = memStat.BuckHashSys
+	m.GaugeMetrics["Frees"] = memStat.Frees
+	m.GaugeMetrics["GCCPUFraction"] = memStat.GCCPUFraction
+	m.GaugeMetrics["GCSys"] = memStat.GCSys
+	m.GaugeMetrics["HeapAlloc"] = memStat.HeapAlloc
+	m.GaugeMetrics["HeapIdle"] = memStat.HeapIdle
+	m.GaugeMetrics["HeapObjects"] = memStat.HeapObjects
+	m.GaugeMetrics["HeapInuse"] = memStat.HeapInuse
+	m.GaugeMetrics["HeapReleased"] = memStat.HeapReleased
+	m.GaugeMetrics["HeapSys"] = memStat.HeapSys
+	m.GaugeMetrics["LastGC"] = memStat.LastGC
+	m.GaugeMetrics["Lookups"] = memStat.Lookups
+	m.GaugeMetrics["MCacheInuse"] = memStat.MCacheInuse
+	m.GaugeMetrics["MCacheSys"] = memStat.MCacheSys
+	m.GaugeMetrics["MSpanInuse"] = memStat.MSpanInuse
+	m.GaugeMetrics["MSpanSys"] = memStat.MSpanSys
+	m.GaugeMetrics["Mallocs"] = memStat.Mallocs
+	m.GaugeMetrics["NextGC"] = memStat.NextGC
+	m.GaugeMetrics["NumForcedGC"] = memStat.NumForcedGC
+	m.GaugeMetrics["NumGC"] = memStat.NumGC
+	m.GaugeMetrics["OtherSys"] = memStat.OtherSys
+	m.GaugeMetrics["PauseTotalNs"] = memStat.PauseTotalNs
+	m.GaugeMetrics["StackInuse"] = memStat.StackInuse
+	m.GaugeMetrics["StackSys"] = memStat.StackSys
+	m.GaugeMetrics["Sys"] = memStat.Sys
+	m.GaugeMetrics["TotalAlloc"] = memStat.TotalAlloc
+	m.GaugeMetrics["RandomValue"] = rand.Float64()
+}
+func (m *MetricsCollector) Send(url string) {
+	for key, val := range m.GaugeMetrics {
+		fullPath := url + "/update/gauge/" + key + "/" + fmt.Sprintf("%v", val)
+		req, err := http.NewRequest(http.MethodPost, fullPath, nil)
+		if err != nil {
+			log.Fatal(err)
+		}
+		client := &http.Client{}
+		res, err := client.Do(req)
+		if err != nil {
+			panic(err)
```

Comment:
Небольшой разрыв соединения с сервером - и агент падает с грохотом:
```
panic: Post "http://localhost:8080/update/gauge/Alloc/275008": dial tcp [::1]:8080: connectex: No connection could be made because the target machine actively refused it.

goroutine 1 [running]:
github.com/timuraipov/alert/internal/agent.(*MetricsCollector).Send(0xc000076380, {0xc0000a9ea0, 0x15})
        C:/work/src/github.com/timuraipov/alert/internal/agent/agent.go:76 +0x406
github.com/timuraipov/alert/internal/agent.(*MetricsCollector).Run(0xc000076380)
        C:/work/src/github.com/timuraipov/alert/internal/agent/agent.go:98 +0x1c5
main.main()
        C:/work/src/github.com/timuraipov/alert/cmd/agent/main.go:13 +0x1b
```